Revise addr space

woodard opened the pull request:
With the latest changes. I feel like it is ready to review. Turning it into a PR also allows people to comment on the patch.

woodard committed:
Begin revising address space proposal

Begin working on revising the address space proposal. This mostly
reorders the text and adds notes for work to be done.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Revise address space proposal

* Change the name of the operator
* Remove implicit pointer stuff
* Remove CFA/CFI stuff deferred until the CFA/CFI proposal.
* Notable rewriting.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Heavy rewrite

Too many changes to describe fully.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Put explicit references to points that need discussion.

Cary's email and Tony's response both made some important points which need
to be discussed in the group as a whole.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Minor cleanups

Fixed a few grammar and spelling mistakes.
Changed the way I presented tables in md

Signed-off-by: Ben Woodard <woodard@redhat.com>

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,369 @@
+>
+>    The DWARF expression is ill-formed if AS is not one of the values
+>    defined by the target architecture specific `DW_ASPACE_*` values.
 
```
IMHO, the relation between `DW_OP_mem` and `DW_OP_addr` should be explained.

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,369 @@
->   attribute that includes a `DW_OP_addr` or `DW_OP_form_tls_address` operator
->   are included; otherwise, they are excluded.
+>    7. `DW_OP_aspace_bregx` ([ULEB128] R, [LEB128] B)
+>        <[integral] AS> → <[location] A >
```
Above we said `[memory location]`, not just `[location]`.

woodard committed:
Respond to Baris's comments

Fix several bugs that Baris pointed out.

Signed-off-by: Ben Woodard <woodard@redhat.com>

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+
+**For Discussion** Is the OpenCL reasoning that Tony cited compelling
+enough that the address space can't be an operand. If so we should
+probably state that in the document.
```
The example is that OpenCL has primitives that take the address space as a runtime value. So I think there needs to be a way to allow computed address space IDs. But there is the very common case that the address is statically known, and that is where an xaddr-like operation that works on the non-default address space could make sense.

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+>    spaces map to target architecture specific DWARF address
+>    spaces. A target architecture may map multiple source language
+>    memory spaces to the same target architecture specific DWARF
+>    address class. Optimization may determine that variable lifetime
```
"address class" -> "address space"

The original proposal did define the separate concepts of address space (for the target architecture implementations) and address classes (for encoding of addresses).

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+>
+>    If AS is an address space that is specific to context elements,
+>    then L corresponds to the location storage associated with the
+>    current context when the location is created, not the context
```
"when the location is created, not the context when that location is used." -> "when the DW_OP_mem operation is evaluated, not the context when that location returned by the evaluation of DW_OP_mem is used."

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+>    size of AS. An object P having the given pointer or reference
+>    type is dereferenced as if the `DW_OP_push_object_location`;
+>    `DW_OP_deref_type` DR; `DW_OP_constu` AS; `DW_OP_mem` expression
+>    was evaluated with the current context except: the result kind is
```
"is" -> "is a"

ccoutant commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+while executing an expression.
+>
+>    The DWARF expression is ill-formed if AS is not one of the values
+>    defined by the target architecture specific `DW_ASPACE_*` values.
```
Suggest: "...is not one of the values defined by the target architecture's ABI."

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+address followed by dereferencing it. Therefore we propose renaming
+those operators to `DW_OP_aspace_deref*` to be consistent with the
+other operators which refer to address spaces and making the old
+`DW_OP_xderef` names aliases.
```
Suggest explicitly mentioning that the necessary changes to support address spaces in CFI will be submitted as a separate issue.

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+address space.
+
+**For discussion** Why isn't this a flavor of addr. We probably need
+to cover relocation here.
```
I added some more detail here and explained those scenarios.

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+>    treating the least significant S bits as an unsigned value A'.
+>
+>    It pushes a memory location within the address space AS whose
+>    offset is A.
```
I don't think that this one is correct. A is the name of the parameter.

woodard replied:
Let's talk about this on Wednesday. I'm not following your train of thought here.

woodard committed:
Address comments from Dec 17 GPU meeting

Address comments and questions that were discussed in the Dec 17 GPU
meeting as well as comments that Tony entered in the PR.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+>    value that represents a target architecture specific address
+>    space identifier AS.
+>
+>    The action is the same as for `DW_OP_breg`, except that R is used
```
I think that was done in the locstack proposal. The current text is:

4. DW_OP_breg0, DW_OP_breg1, . . . , DW_OP_breg31

The single operand of the DW_OP_breg<n> operations provides a signed
LEB128 byte offset. The contents of the specified register (0-31) are treated as
a memory address **in the default address space**. The offset is added to the
address obtained from the register and the resulting memory location is
pushed onto the stack.

I think that is what you are looking for. If not please comment further.

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,372 @@
+>    `DW_OP_deref_type` DR; `DW_OP_constu` AS; `DW_OP_mem` expression
+>    was evaluated with the current context except: the result kind is
+>    location description; the initial stack is empty; and the object
+>    is the location of P.
```
From the preceding text

>    DR is the offset
>    of a hypothetical debug information entry D in the current
>    compilation unit for an integral base type matching the address
>    size of AS.

(If I wrote that, I don't recall writing it. It doesn't feel like my style.")  

I'll take a stab at making it simpler.  Check back when I push my next commit.

woodard committed:
Additional changes from the PR/meeting

I missed a few later breaking comments in the PR. Added those changes.

Signed-off-by: Ben Woodard <woodard@redhat.com>

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -305,15 +320,15 @@ the following paragraph:
+>    omitted, this defaults to `DW_ASPACE_default`. An object P having
+>    the given pointer or reference type is dereferenced as if the
+>    `DW_OP_push_object_location`; `DW_OP_deref_type` referencing is
+>    the offset of a hypothetical DIE in the current compilation unit
```
"referencing is the" -> "referencing the"

woodard committed:
Fix a typo that tony found.

Signed-off-by: Ben Woodard <woodard@redhat.com>

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,388 @@
+0x101 are the bits immediately after the ones in address
+0x100. However, some GPU memory pools are designed such that when
+accessed a particular way the bits at address 0x101 are a stride away
+from the bits at 0x100 that way a particular compute unit can address
```
"that way" -> "such that"

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,388 @@
+This stride addressed memory pool may reference the same bank of bits
+may be accessible from a different address space a different way, that
+is fine. There is no requirement that there is only one way to access
+a particular set of bits. In fact, it is common for a GPU to provide
```
"This stride addressed memory pool may reference the same bank of bits as can be accessible from a different address space using a different embedded access pattern. There is no requirement that there is only one way to access a particular set of bits."

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,388 @@
+is fine. There is no requirement that there is only one way to access
+a particular set of bits. In fact, it is common for a GPU to provide
+access to the same bits through different address spaces. One access
+pattern is used for serial code while another is used vectorized code.
```
"used" -> "used for"

t-tye reviewed (changes requested):
Read through and suggested a collection of changes, but overall I really like this:-)

woodard committed:
Address many of Tony's comments.

Tony did a careful read through and picked up some minor errors. I
fixed most of them. Only two are left, I believe that both of those
require discussion.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'upstream' into revise-addr-space

woodard committed:
Changes before GPU meeting

Fix action item before meeting.
Add some notes to a problematic paragraph.

Signed-off-by: Ben Woodard <woodard@redhat.com>

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+therefore context independent memory. GPU memory is often context
+dependent. The definition of this context has already been accepted
+into the DWARF6 standard with issue 241011.1 Expression Evaluation
+Context [241011.1](https://dwarfstd.org/issues/241011.1.html). Unlike
```
Nit: Issue number is duplicated.

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+system memory where an address like 0x1000 refers to the same location
+independent of the evaluation context. GPU memory pools can be local
+to a GPU's processing unit and every processing unit may have its own
+memory pool. Thus, within a specific address space, the address 0x1000
```
We can give concrete examples from vendors. E.g. On Intel GPUs, there is shared local memory for each Xe core: https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2025-2/shared-local-memory.html
I found the following corresponding documentation for AMD: https://rocm.docs.amd.com/projects/HIP/en/develop/understand/hardware_implementation.html#local-data-share-lds
NVIDIA calls the same thing "shared memory".

woodard replied:
Let's see if the nvidia guys feel left out and contribute their own link. ;-)

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+expressions which conceptually undoes the work done by the addressing
+mode in order to point the consumer to the correct location. This
+would expand the size of the DWARF generated and it would make the job
+of generating debuginfo harder for the producer.
```
These 3 paragraphs are IMHO not crucial for motivating address spaces but they incur a substantial mental load on the reader.  Above we say GPUs have address spaces (I proposed we give concrete vendor examples) and that should be enough motivation. I'd suggest we remove these 3 paragraphs for simplifying the proposal.

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+DWARF 5 has the concept of an address in many expressions but does not
+define how it relates to address spaces. For example,
+`DW_OP_push_object_location` pushes the address of an object. Other
+contexts implicitly push an address on the stack before evaluating an
```
implicitly push a *location* on the stack

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+addresses in the system memory. That would require the consumer to
+have special treatment for such values. Furthermore, purely swizzled
+pointers do not capture the semantic differences of address spaces
+like context dependence, pointer size, or embedded access patterns.
```
For this section, I have the same opinion I had above: it incurs unnecessary mental load on the reader while being not crucial for explaining why we need address space support in DWARF. I propose that we remove this section. If someone from the committee asks about swizzling during the discussion, it can be explained, but having this section derails the flow.

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+>    then the pushed location L corresponds to the location storage
+>    associated with the current context when the `DW_OP_mem`
+>    operation is evaluated, not the context when the location
+>    returned by the evaluation of `DW_OP_mem` is used.
```
This implicitly makes a memory storage keep a copy of its context. How does a consumer know which context to save in a memory storage?

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+>    attribute with a constant value AS representing an architecture
+>    specific DWARF address space (see 2.12 "Address Spaces"). If
+>    omitted, this defaults to `DW_ASPACE_default`. An object P having
+>    the given pointer or reference type is dereferenced as if the
```
I think what's meant by "dereferencing" here is the C-style `*p` operation, which means a double-read in DWARF ops.  So, I was expecting to see another `DW_OP_deref` after `DW_OP_mem` below. If what's meant by "dereferencing" is a single read, then I was expecting just `DW_OP_push_object_location`; `DW_OP_deref_type` and not the rest with `DW_OP_constu AS; DW_OP_mem`. So, a single read results in a value.

woodard committed:
Address Baris's comments

Baris pointed out several problems on the PR. This attempts to address
most of them.

Signed-off-by: Ben Woodard <woodard@redhat.com>

barisaktemur commented on 013-generalize-address-spaces.md:
```diff
@@ -11,48 +11,41 @@ differently.
+one access pattern for serial code while another is used for
+vectorized code. If this access pattern were not conceptually
+abstracted into an address space, then the compiler would have to
+generate DWARF expressions which conceptually undoes the work done by
```
"undoes" -> "undo"

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -11,48 +11,41 @@ differently.
-mode in order to point the consumer to the correct location. This
-would expand the size of the DWARF generated and it would make the job
-of generating debuginfo harder for the producer.
+memory pool. A couple of examples of this are: [Intel's SharedLocal
```
The rendering of the font in my browser makes it look like it is supposed to be SharedLocal vs. Shared Local

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -11,48 +11,41 @@ differently.
+the addressing mode in order to point the consumer to the correct
+location. This would expand the size of the DWARF generated and it
+would make the job of generating debuginfo harder for the
+producer. Another common example of an embedded access pattern is
```
disagree. 
Part of what I did there to condense it was describe serial access mode vs. wave access modes as an embedded access pattern. This would be the case where a SIMT processor uses the vector access mode in a subgroup or wave's thread execution but then spills the whole vector register to memory with a store or move. Previously I wasn't lumping this under an access mode.  Following that with "Another common example" seems to sequence it. If strided memory were the first then I would use that and then describe wave/subgroup access modes following "Another common example"

woodard committed:
Mostly fix the problematic paragraph.

1. This fixes a couple of minor bugs that Baris found.
2. It also rearranges where I describe the problems with xderef
3. However, the biggest change is rewriting the problematic paragraph
describing DW_AT_address_space/

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,388 @@
+>    `DW_OP_constu` AS; `DW_OP_mem` expression was evaluated with the
+>    current context except: the result kind is a location
+>    description; the initial stack is empty; and the object is the
+>    location of P.
```
please take a look at the latest version.

woodard committed:
Fix typos

Signed-off-by: Ben Woodard <woodard@redhat.com>

t-tye commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+
+DWARF 5 has the concept of an address in many expressions but does not
+define how it relates to address spaces. For example,
+`DW_OP_push_object_location` pushes the address of an object. Other
```
I think this is discussing DWARF 5 which only allows address values on the stack. Not until DWARF 6 do we get locations on the stack.

woodard committed:
Merge branch 'upstream' into revise-addr-space

woodard commented on 013-generalize-address-spaces.md:
```diff
@@ -2,429 +2,419 @@
+`DW_OP_mem` can be a thought of as a more general form of `DW_OP_addr`
+with `DW_OP_addr` being a short hand form of:
+
+	DW_OP_lit0 ; DW_AS_default is by definition 0
```
Disagree we are not compilers we are humans and can deal better with forward declarations. That is why I put it in a comment.

woodard committed:
Address many of Baris's latest suggestions

Baris provided additional feedback and I applied many of his suggested
changes.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Reword definition of mem

As discussed in the meeting, the algorithmic description about how the
size changes are handled when A doesn't match the size of the address
space is replaced with simpler text.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Final updates

Applied all of Baris's suggestions from his final review.
Fixed the verb agreement with undo vs. undoes in the introductory text.
Adjusted the description of DW_OP_mem to address Tony's concern.

Signed-off-by: Ben Woodard <woodard@redhat.com>
